feat(sync): RN-1545: Epic datatrak sync

chris-bes opened the pull request:
## RN-1545: Epic datatrak sync

> [!IMPORTANT]
> ### Manual release steps
> 
> - [x] Before kick-off, update **deployment** Lambda with latest code from epic branch
> - [x] Replace the `schema_change_notification` function
>
>```
>CREATE OR REPLACE FUNCTION public.schema_change_notification() RETURNS event_trigger
>    LANGUAGE plpgsql
>    AS $$
>  DECLARE
>    obj RECORD;
>    dominated_by_ignored BOOLEAN := TRUE;
>  BEGIN
>  FOR obj IN SELECT * FROM pg_event_trigger_ddl_commands()
>  LOOP
>    IF obj.schema_name != 'sync_snapshots' AND NOT obj.in_extension AND
>       NOT EXISTS (
>         SELECT 1 FROM pg_class c
>         WHERE c.oid = obj.objid AND c.relpersistence = 't'
>       )
>    THEN
>      dominated_by_ignored := FALSE;
>      EXIT;
>    END IF;
>  END LOOP;
>
>  IF NOT dominated_by_ignored THEN
>    PERFORM pg_notify('schema_change', 'schema_change');
>  END IF;
>  END;
>  $$;
>```

## 🥞 Stack

1. #6289 👈 
2. #6446

jaskfla commented on packages/constants/package.json:
```diff
@@ -0,0 +1,28 @@
+    "hoistingLimits": "workspaces"
+  },
+  "scripts": {
+    "build": "rimraf dist && npm run --prefix ../../ package:build:ts",
+    "build-dev": "yarn run build",
+    "build-watch": "rimraf dist && npm run --prefix ../../ package:build:ts-watch",
```
Occurred to me this won’t be caught by Git conflict resolution:

```suggestion
    "build": "rimraf dist && yarn run package:build:ts",
    "build-dev": "yarn run build",
    "build-watch": "rimraf dist && yarn run package:build:ts-watch",
```

jaskfla commented on packages/datatrak-web/src/features/CountrySelector/useUserCountries.ts:
```diff
@@ -1,6 +1,8 @@
 import { ChangeEvent, ChangeEventHandler, useState } from 'react';
 
-import { Country } from '@tupaia/types';
+import { camelcaseKeys } from '@tupaia/tsutils';
```
So sorry, I know I reviewed this but I just caught this! We’ve [historically had issues](https://beyondessential.slack.com/archives/C032DCHKTGR/p1716764370266429) using **@tupaia/tsutils** as a dependency of front-end packages because it has some transitive dependencies that only work in Node

Since this PR does some work to make DataTrak support some server-y code, I’m not sure if this issue will still apply

If you run into `Uncaught ReferenceError: process is not defined` after deploying (not locally), then you might need to pull `camelcaseKeys` into **@tupaia/datatrak-web**. (If it works find, remember to declare the dependency! It’s currently missing)

jaskfla commented:
Looks like **@tupaia/datatrak-web** is missing these dependencies:

- `case` (used in `./src/sync/insertSnapshotRecords.ts`), and
- `knex` (dev dependency used in `./src/database/DatatrakDatabase.ts`)

jaskfla commented on packages/central-server/src/dataAccessors/answerBodyParsers.js:
```diff
@@ -0,0 +1,61 @@
+
+  if (isValidFileId(answer.body)) {
+    const s3ImagePath = getS3ImageFilePath();
+    return `${S3_BUCKET_PATH}${s3ImagePath}${answer.body}.png`;
```
e7d2e40b1aafa1bbb694fc012d8d5a9280ba53cf

jaskfla commented on packages/database/.babelrc.js:
```diff
@@ -21,7 +40,7 @@ const getIgnore = api => {
 
         if (directory === 'migrations' || directory === 'migrationData') {
-          return checkMigrationOutdated(filename);
+          return checkMigrationOutdated(filename, filepath);
```
This has been addressed in 584de50135481d57ead57656aed364d12b0e07eb, which is part of #6446

jaskfla committed:
refactor(database): make `sync_session.info` JSONB

jaskfla committed:
tweak: various type hints

jaskfla committed:
tweak(centralServer): short circuit `EditUserAccounts`

jaskfla committed:
tweak(centralServer): remove unused import

jaskfla committed:
refactor: dedupilcate `snakeKeys` function

chris-bes committed:
add columnToMerge

chris-bes committed:
remove null

chris-bes committed:
clean up if stream throws error

jaskfla committed:
tweak(datatrakWeb): tighten `ClientSyncManager` types

jaskfla committed:
tweak: tighten types

jaskfla committed:
tweak(centralServer): remove unused import

jaskfla committed:
fix(sync): bad type

jaskfla committed:
tweak: type database `recordType` (and similar) symbols

jaskfla committed:
refactor(syncServer): don’t build unused map

jaskfla committed:
refactor(datatrakWeb): use safe bindings in `clearDatabase`

jaskfla committed:
tweak(database): more survey type hints

jaskfla committed:
fix(database): bad types

jaskfla committed:
refactor(database): use safe binding in `Survey` model

jaskfla committed:
tweak(database): type `FeedItem.createAccessPolicyQueryClause`

jaskfla committed:
style(datatrak): unify `FeedItem` and `Survey` model code style

jaskfla committed:
refactor(database): fetch only necessary columns

jaskfla committed:
tweak(database): organise exports

chris-bes committed:
missing await

chris-bes committed:
save more memory

chris-bes committed:
remove

jaskfla committed:
tweak(centralServer): declare `permissionsFilteredInternally` overrides as `const`s

jaskfla committed:
revert(database): restore missing exported types

This reverts commit b7377af808207678716c3f22f3f95c3d9f62bea9.

jaskfla committed:
test(database): work around Jest’s ES6 module limitation

chris-bes committed:
fix test

jaskfla committed:
test(database): fix same error with `SCHEMA`

jaskfla committed:
doc(sync): fix bad comment

chris-bes committed:
add some logs for debug

chris-bes committed:
add initial memory

chris-bes committed:
log error

chris-bes committed:
revert

chris-bes committed:
add more debug

jaskfla committed:
tweak(database): `deleteById` return type hint

jaskfla committed:
tweak(dabase): define model `join`s as `const`s

jaskfla committed:
tweak(database): narrower model types

jaskfla committed:
style(database): organise imports

jaskfla committed:
tweak(centralServer): throw nongeneric error

jaskfla committed:
doc(datatrakWeb): remove dead comment

chris-bes committed:
change batch size

chris-bes committed:
reduce batch size

chris-bes committed:
add more debug logs

chris-bes committed:
more logging

chris-bes committed:
fix batching

chris-bes committed:
more debug logs

chris-bes committed:
fixed debug

chris-bes committed:
more debug

chris-bes committed:
more debug

chris-bes committed:
update batch size

chris-bes committed:
more debug

chris-bes committed:
fix debug log

chris-bes committed:
updated batch size

jaskfla committed:
test: use valid Base64 URI in relevant unit tests

jaskfla committed:
test(centralServer): replace magic strings with enums

jaskfla committed:
fix(datatrak): RN-1752: `File` answer text format from offline uploads (#6623)

* neater typing

* shallower indenting

* const

* fix

* lint

* more readable

* remove dead comment

* lint

chris-bes committed:
fix batching

jaskfla committed:
tweak(datatrakWeb): narrower return type

jaskfla committed:
tweak(datatrakWeb): remove MediTrak reference from home view

jaskfla committed:
fix(datatrak): RN-1491: filter local surveys by both access policy and country (#6622)

* builds

* dedupe `ComparisonType` definition

* move existing logic to model layer

fr

* fix bad param

* format SQL

* fetch only necessary column

* fix

* typo

* types

* remove verbose comments

* types

* remove redundant error message definition

* fix use of incorrect models

* dedupe country ID clause

* remove dead import

* lazy typing

* safer

* use `orThrow`

* less blocking

* use `import type`

* more readable

jaskfla committed:
tweak: fix unsafe typing

chris-bes committed:
increase memory 2nd attempt

chris-bes committed:
increase

chris-bes committed:
increase memory

chris-bes committed:
vacuum

chris-bes committed:
create only

chris-bes committed:
no vacuum

chris-bes committed:
nginx + log

chris-bes committed:
more logs

chris-bes committed:
relaxedDurability

chris-bes committed:
upgrade

chris-bes committed:
test remove relaxedDurability

jaskfla committed:
Merge branch 'dev' into rn-1545-epic-datatrak-offline

# Conflicts:
#	packages/database/src/core/modelClasses/Project.js
#	packages/database/src/modelClasses/User.js
#	packages/datatrak-web-server/src/routes/SurveyRoute.ts
#	packages/datatrak-web-server/src/routes/UserRoute.ts
#	packages/web-config-server/src/apiV1/projects.js

jaskfla committed:
merge: fix bad merge resolution

jaskfla committed:
refactor(database): use `exists` method

jaskfla committed:
tweak(centralServer): narrow `UserRecord` types

jaskfla committed:
Merge branch 'rn-1545-epic-datatrak-offline' into rn-1717-test-four

# Conflicts:
#	packages/database/src/core/DatabaseModel.js
#	packages/database/src/core/modelClasses/Project.js
#	packages/database/src/core/modelClasses/User/User.js
#	packages/datatrak-web-server/src/routes/UserRoute.ts

jaskfla committed:
fix(database): bad `database.connection.raw` usage

jaskfla committed:
Merge branch 'rn-1545-epic-datatrak-offline' into rn-1717-test-four

jaskfla committed:
refactor(webConfigServer): more efficient `fetchHasPendingProjectAccess`

chris-bes committed:
revert unnecessary logs

chris-bes committed:
add logs for memory for incremental sync

chris-bes committed:
optional

jaskfla committed:
fix(database): restore missing type

jaskfla committed:
style(database): reflow awkward inline comment

chris-bes committed:
increase timeout

chris-bes committed:
increase timeout

chris-bes committed:
remove timeout

jaskfla committed:
fix(database): `debug_log` update race condition

chris-bes committed:
revert nginx

chris-bes committed:
add debug log

chris-bes committed:
add log

chris-bes committed:
remove users

jaskfla committed:
tweak(entityServer): narrower types

jaskfla committed:
fix(serverBoilerplate): unsafe property access

jaskfla committed:
refactor(entityServer): improve static type inference

jaskfla committed:
tweak(database): improve Task model & record type hints

jaskfla committed:
tweak(datatrak): clearer errors in bad mutation function calls

jaskfla committed:
tweak: custom error with unsafe accesses

jaskfla committed:
refactor(entityServer): fix unsafe access, short circuit

jaskfla committed:
tweak(tsmodels): bolster `Model` generic typing

jaskfla committed:
fix(datatrakWebServer): fetching a survey’s project

jaskfla committed:
tweak(datatrakWeb): fix query key collision

jaskfla committed:
tweak(serverUtils): use `node:` protocol import

jaskfla committed:
tweak(database): sync-related model instance method type hints

jaskfla committed:
tweak(constants): tighten `const` types

jaskfla committed:
refactor(datatrakWebServer): await promise

jaskfla committed:
style(datatrakWebServer): remove useless template

jaskfla committed:
refactor(datatrakWeb): more idiomatic query function

jaskfla committed:
fix(datatrak): RN-1490: make survey submission history available offline (#6625)

* tweak: fix unsafe typing

* fix bad type

* move to model layer

* use offline

* clean up diff

* alphasort

update schema

* more explicit

* better query key

* reuse object

* format

* group by primary key

* ensure

* fix

* again

jaskfla committed:
fix(datatrakWebServer): bad type

jaskfla committed:
tweak(datatrakWebServer): 404 when task not found

jaskfla committed:
tweak(datatrakWebServer): use const, interface

jaskfla committed:
refactor(datatrakWebServer): parallelise awaits

jaskfla committed:
fix(datatrakWeb): capitalisation of DataTrak

jaskfla committed:
tweak(datatrakWeb): refine update modal typography

jaskfla committed:
refactor: faster `ProjectModel.getAccessibleProjects` (marginally)

jaskfla committed:
tweak(tsmodels): tighten `MultiJoinItem` types

jaskfla committed:
tweak(centralServer): use `const`

jaskfla committed:
refactor(centralServer): short circuit if permission assertion fails

jaskfla committed:
refactor(centralServer): join in SQL

jaskfla committed:
refactor(datatrakWebServer): use `const`, `interface`

jaskfla committed:
tweak(datatrakWeb): improve query key

jaskfla committed:
refactor(datatrakWebServer): dedupe constant definition

jaskfla committed:
format(database): reflow SQL string

jaskfla committed:
tweak(database): more comprehensive SyncLookupQueryDetailsBuilder

jaskfla committed:
fix(tsmodels): bad type

jaskfla committed:
tweak(types): tighten `FeedItem` types

jaskfla committed:
format(centralServer): enforce code format

jaskfla committed:
style(centralServer): lint `importSurveyResponses`

chris-bes committed:
sync all entities

jaskfla committed:
refactor(dhisApi): use set semantics for cacheable endpoints

jaskfla committed:
refactor(centralServer): use `findOneOrThrow`, await promises

jaskfla committed:
tweak(centralServer): faster check for existing account

jaskfla committed:
tweak(centralServer): `emailVerification` type hints

jaskfla committed:
tweak(centralServer): `409 Conflict` when account exists

jaskfla committed:
refactor(centralServer): less I/O marking user as verified

jaskfla committed:
fix: hyphenation of ‘email’

jaskfla committed:
tweak(database): remove unused `@typedef`

jaskfla committed:
Merge branch 'dev' into rn-1545-epic-datatrak-offline

# Conflicts:
#	packages/database/src/core/modelClasses/Project.js

jaskfla committed:
Merge branch 'rn-1545-epic-datatrak-offline' into rn-1717-test-four

jaskfla committed:
refactor(database): simplify `getAllProjectDetails` binding

jaskfla committed:
refactor(sync): faster save phase (#6604)

* use `partition`

* simplify detection of existing records

* fetch only ID

* one fewer iteration

* fix ambiguous projection

* more readable

* cleaner diff

* remove useless negation

* fix(database): `debug_log` update race condition

* use explicit cast

jaskfla committed:
Merge branch 'dev' into rn-1545-epic-datatrak-offline

jaskfla committed:
Merge branch 'rn-1545-epic-datatrak-offline' into rn-1717-test-four

chris-bes committed:
Revert "sync all entities"

This reverts commit d074f8de078103195904893ab20bc1531b56101c.

chris-bes committed:
sync all entities

chris-bes committed:
clean up EntityHierarchyCacher

jaskfla committed:
refactor(centralServer): avoid needless fetch

jaskfla committed:
tweak(constants): more permission group name consts

jaskfla committed:
tweak(constants): dedupe object ID regex pattern

chris-bes committed:
repeatable read

jaskfla committed:
refactor(database): avoid needless `Date` instantiation

chris-bes committed:
remove repeatable read

chris-bes committed:
fix advisory lock

jaskfla committed:
refactor(centralServer): remove redundant predicate

jaskfla committed:
tweak(datatrakWeb): educate apostrophe

chris-bes committed:
Revert "fix advisory lock"

This reverts commit 962795ddf260e2ccd5ce998839495c2bb4f8e7ee.

chris-bes committed:
add autobind to tupaia database

jaskfla committed:
fix(database): bad `getTimezone` type and usages (#6637)

jaskfla committed:
refactor(datatrakWeb): set semantics for read-only question types

jaskfla committed:
refactor(datatrakWeb): extract survey route matchers, dedupe

jaskfla committed:
chore: dedupe type guards, delete unused

jaskfla committed:
refactor(centralSurvey): fetch only necessary column

jaskfla committed:
style(sync): remove useless templates

jaskfla committed:
tweak(apiClient): use const type

jaskfla committed:
format(datatrak): enforce code format

chris-bes committed:
fix(sync): RN-1717: Fix advisory lock causing db connection pool exhaustion (#6638)

* fix advisory lock

* comment

* address reviews

jaskfla committed:
refactor(datatrakWeb): faster check for unsynced data

jaskfla committed:
refactor(datatrakWeb): simplify project button label

jaskfla committed:
refactor(database): delegate ObjectID timestamp to bson-objectid

jaskfla committed:
tweak(database): `generateId` type hints

jaskfla committed:
fix(centralSever): RN-1711: slow email verification (#6634)

* tweak(centralServer): new accounts first when verifying email

* more readable

jaskfla committed:
fix(datatrakWeb): RN-1490: offline survey response history (#6632)

* move to model layer

* format(database): enforce code format

* tighten types

* bring offline

* dedupe

* remove dead import

* oops

* 🩹

* fetch only necessary data

* add missing sort

* fix

* more consistent

* add missing LIMIT clause

jaskfla committed:
chore(database): remove unused symbols

jaskfla committed:
style(database): no assignment in expression

jaskfla committed:
chore(database): dedupe instance method

jaskfla committed:
chore: remove suspicious `forEach` callback return

jaskfla committed:
refactor: avoid useless `Date` instance

jaskfla committed:
refactor(database): use optional chain

jaskfla committed:
refactor(database): use literal key

jaskfla committed:
refactor(database): use `indexOf`

jaskfla committed:
chore(utils): remove unused symbols

jaskfla committed:
refactor(utils): faster `calculateOuterBounds`

jaskfla committed:
chore: remove unused imports

jaskfla committed:
style: misleading variable name

chris-bes committed:
clean up properly after migration

jaskfla committed:
fix(database): undefined `handlerLock` in `TupaiaDatabase` subclass instances (#6641)

* fix

* comment

jaskfla committed:
refactor(datatrakWeb): define constants at top-level scope

jaskfla committed:
format(datatrakWeb): enforce code format

jaskfla committed:
format(datatrakWeb): reflow long comment

jaskfla committed:
tweak(datatrakWeb): use transaction to get user locally

jaskfla committed:
refactor(datatrakWeb):  use `Object.hasOwn`

jaskfla committed:
tweak(datatrakWeb): safer default local mutation function context type

jaskfla committed:
chore(datatrakWeb): use literal key

jaskfla committed:
chore(datatrakWeb): remove suspicious `let` declaration

jaskfla committed:
style(datatrakWeb): use optional chain

jaskfla committed:
chore(datatrakWeb): remove useless fragment

jaskfla committed:
refactor(datatrakWeb): remove spread from reducer

jaskfla committed:
refactor(datatrak): fix dubious activity feed Effect dependencies

jaskfla committed:
refactor(datatrakWeb): use only relevant primitive as dependency

jaskfla committed:
tweak(datatrakWeb): use transaction to save user locally

jaskfla committed:
tweak(datatrakWeb): remove bad selector

jaskfla committed:
tweak(datatrakWeb): clearer error message

jaskfla committed:
tweak(datatrakWeb): get entity descendants with transaction

jaskfla committed:
tweak(apiClient): use const type

jaskfla committed:
fmt(apiClient): enforce code format

jaskfla committed:
tweak(datatrakWeb): narrower inferred type

chris-bes committed:
fix issue sometimes there's no data showing after initial sync

chris-bes committed:
revert all debug logs

chris-bes committed:
revert debug logs 2

chris-bes committed:
remove logs

jaskfla committed:
tweak(apiClient): use const

chris-bes committed:
clean up datatrak auto update

jaskfla committed:
fmt(syncServer): format `createApp.ts`

jaskfla committed:
tweak(datatrakWeb): RN-1806: show loading spinner while logging out (#6643)

chris-bes committed:
clean up update

jaskfla committed:
refactor(datatrakWeb): use `find…OrThrow` helper methods

jaskfla committed:
tweak(uiComponents): define spinner once, balance message

chris-bes committed:
Merge branch 'dev' into rn-1545-epic-datatrak-offline

jaskfla committed:
fix(centralServer): unsafe property access

jaskfla committed:
tweak: more readable lists in error messages

jaskfla committed:
tweak(centralServer): annotate abstract methods

jaskfla committed:
chore(centralServer): lint unused symbols

jaskfla committed:
refactor(centralServer): use optional chain

jaskfla committed:
test(serverUtils): update unit test to reflect expected behaviour

jaskfla committed:
fix(datatrak): public survey submission

jaskfla committed:
tweak(datatrakWeb): clearer useLogin onSuccess error message

jaskfla committed:
tweak(datatrakWeb): memoise `SyncContext` value

jaskfla committed:
refactor(datatrak): decompose sync status listener hooks (#6585)

* type emitter

* extract into separate hooks

* convenience wrapper

* use new hook

* read `isSyncing` from `syncManager` every time

* refactor

* preserve existing behaviour

jaskfla committed:
fix(datatrakWeb): RN-1808: refreshing on sync page (#6651)

* type emitter

* extract into separate hooks

* convenience wrapper

* use new hook

* read `isSyncing` from `syncManager` every time

* refactor

* preserve existing behaviour

* fix

oops

* memoise event handler

* organise imports

* extract hook

jaskfla committed:
feat(datatrak): RN-1751: require up-to-date client to sync (#6644)

* enforce code format

* kinda sorta mostly there

* fixes

* update `attachApiClientToContext` interface

* use global axios `timeout`

fix bad export

* remove redundant custom handling

* refactors

* attach `X-Client-Version` to all Axios requests

* expore `X-Required-Client-Version` header

* display client version in GUI

* update comment

* remove debug logs

* refactor

* simplify

* delegate throw to backend

* simplify

* remove dead error handling

* remove redundant helper

* use semver

* un-propose `X-Required-Client-Version` header

* comment middleware side effect

* use CSS comment syntax

* trailing full stop

* test(syncServer): RN-1751: version compatibility test suite (#6647)

* try

* fix bad mock

* simplify

* delete stub test

* use `it.each`

* reorder

* refactor

* comment unrealistic cases

* fix

* typo

* proper reset before each

* mock next fn

fix

* less verbose

* dedupe type casts

* just mock it

* remove exclusive test

* fix test descriptions

* fix bogus `main` entrypoint

* update test mocks

* show app version text only in web app

jaskfla committed:
doc: list @tupaia/sync as a library package

chris-bes committed:
fix bug when mark record as deleted but record does not exist yet

chris-bes committed:
feat(datatrak): RN-1795: Notify user to reset data when permissions are changed (#6616)

* notify reset data

* use unknown

* clean up

* remove unused

* clean up

* clean up

* clean up

* add comment

* clean up

* clean up

* remove unused import

* add MIGRATIONS record type

* fix bad merge resolution

* type `countSyncSnapshotRecords` `parameters`

* fix build

* code format

* organise imports

* prefer interface

* lint Effect

* update wording for UX clarity

* `color-scheme` on banner

* code style

* fix hook rule violation

* code style

* smarter colouring

* extract

* add missing(?) mock

* fix browser login

* clearer errors

* refactor: avoid shadowing

* await in logout onSuccess

* use const

* types

* slightly faster logout

* move redirect into logout mutation fn

* fewer queries

* remove dead import

* fixes and tidy-ups

* refactor(datatrakWeb): simplify project button label

* faster permission change check

* this works

simplify

* improve type

* better wording

* Squashed commit of the following:

commit c7fe7fab59f3875d1151f2875d29fdbff378115e
Author: Jasper Lai <33956381+jaskfla@users.noreply.github.com>
Date:   Mon Feb 16 18:23:11 2026 +1300

    type

commit 9668770c0ffb5ef2f74efb5fb2d94186e5c1467f
Author: Jasper Lai <33956381+jaskfla@users.noreply.github.com>
Date:   Mon Feb 16 17:40:30 2026 +1300

    refactor

    simplify

    fix

    cherry pick

* Squashed commit of the following:

commit 37a04b8dd40570995d63d1bed8e83d2796d400ef
Author: Jasper Lai <33956381+jaskfla@users.noreply.github.com>
Date:   Mon Feb 16 21:11:41 2026 +1300

    extract

    # Conflicts:
    #	packages/datatrak-web/src/components/ConfirmationModal.tsx
    #	packages/datatrak-web/src/hooks/useAbandonSurveyGuard.tsx

commit 50082806e4557102968e5d7d841134af77148172
Author: Jasper Lai <33956381+jaskfla@users.noreply.github.com>
Date:   Mon Feb 16 18:35:07 2026 +1300

    fix

commit c7fe7fab59f3875d1151f2875d29fdbff378115e
Author: Jasper Lai <33956381+jaskfla@users.noreply.github.com>
Date:   Mon Feb 16 18:23:11 2026 +1300

    type

commit 9668770c0ffb5ef2f74efb5fb2d94186e5c1467f
Author: Jasper Lai <33956381+jaskfla@users.noreply.github.com>
Date:   Mon Feb 16 17:40:30 2026 +1300

    refactor

    simplify

    fix

    cherry pick

* better prop names

* use logical props

* tighten types

* better naming

* fix

partial fix

redirect not working

* not quite

* use styled-components

* simplify and dedupe

remove dead import

* fix misleading symbol name

* clearer error message

* explicit refetch

* less cryptic error message

* tweak(datatrakWeb): use transaction to save user locally

* show only when online

* get permissionChanged with useDatabaseQuery

* clearer error message

* tweak(apiClient): use const

* fix bad resolution

* fix bad refetches after logout

* more readable

* feat(datatrak): RN-1608: Update notification UI change (#6646)

* update notification ui change

* fix color

* fix build

---------

Co-authored-by: Jasper Lai <33956381+jaskfla@users.noreply.github.com>

chris-bes committed:
fix reset data

chris-bes committed:
remove debug log in stream

chris-bes committed:
remove unused

chris-bes committed:
check if sync lookup is ready before allowing sync

chris-bes committed:
remove log

chris-bes committed:
export logMemory

chris-bes committed:
update yarn.lock

chris-bes committed:
fix test

chris-bes committed:
fix error

chris-bes committed:
fix test

chris-bes committed:
Merge pull request #6446 from beyondessential/rn-1717-test-four

feat(sync): RN-1717: Sync bugs

chris-bes committed:
rename to move migrations to more recent

chris-bes committed:
add comment
